get_password_darwin_test.go

```diff
@@ -0,0 +1,97 @@
+// +build darwin
+
+package sshvault
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+	"syscall"
+	"testing"
+
+	"github.com/keybase/go-keychain"
+	"github.com/kr/pty"
+)
+
+func InjectKeychainPassword(path, pw string) error {
+	item := keychain.NewItem()
+	item.SetSecClass(keychain.SecClassGenericPassword)
+	item.SetLabel(fmt.Sprintf("SSH: %s", path))
+	item.SetService("SSH")
+	item.SetAccount(path)
+	item.SetData([]byte(pw))
+	item.SetSynchronizable(keychain.SynchronizableNo)
+
+	return keychain.AddItem(item)
+}
+
+func DeleteKeychainPassword(path string) error {
+	item := keychain.NewItem()
+	item.SetSecClass(keychain.SecClassGenericPassword)
+	item.SetService("SSH")
+	item.SetAccount(path)
+
+	return keychain.DeleteItem(item)
+}
+
+func TestKeychain(t *testing.T) {
+	key_pw := "argle-bargle"
+	key_bad_pw := "totally-bogus\n"
+
+	dir, err := ioutil.TempDir("", "vault")
+	if err != nil {
+		t.Error(err)
+	}
+	defer os.RemoveAll(dir) // clean up
+
+	tmpfile := filepath.Join(dir, "vault")
+
+	vault, err := New("test_data/id_rsa.pub", "", "create", tmpfile)
+	if err != nil {
+		t.Error(err)
+	}
+	key_path, err := filepath.Abs(vault.key)
+	if err != nil {
+		t.Errorf("Error finding private key: %s", err)
+	}
+	err = InjectKeychainPassword(key_path, key_pw)
+	if err != nil {
+		t.Errorf("Error setting up keychain for testing: %s", err)
+	}
+	defer DeleteKeychainPassword(key_path) // clean up
+
+	_, tty, err := pty.Open()
+	if err != nil {
+		t.Errorf("Unable to open pty: %s", err)
+	}
+
+	// File Descriptor magic. GetPasswordPrompt() reads the password
+	// from stdin. For the test, we save stdin to a spare FD,
+	// point stdin at the file, run the system under test, and
+	// finally restore the original stdin
+	old_stdin, _ := syscall.Dup(int(syscall.Stdin))
+	old_stdout, _ := syscall.Dup(int(syscall.Stdout))
+	syscall.Dup2(int(tty.Fd()), int(syscall.Stdin))
+	syscall.Dup2(int(tty.Fd()), int(syscall.Stdout))
+
+	go PtyWriteback(pty, key_bad_pw)
```

Comment:
Whoops, sorry, I should have moved its definition into the test helper file. The function is in `vault_test.go`

It's an asynchronous write to the the PTY that's being used to fake stdin for the prompter. The prompter function disables echoing back anything typed, which involves calling ioctl on the PTY. Further, the prompter's read is blocking IO. Writing to the PTY before invoking the prompter doesn't work, so I used a goroutine to do the write async. If you check `PtyWriteback`, there's a 500ms sleep to allow the prompter to start its blocking read.

Yea, it's a race condition. But a reliable one, this test is already fairly complex, and I didn't want to start adding proper locking on top.